src/main/kotlin/com/exactpro/th2/common/utils/message/MessageHolder.kt

```diff
@@ -0,0 +1,195 @@
+                            }
+                            it.listValue.getValues(index)
+                        }
+
+                        else -> error("Field '$name' can't be got from unknown value: ${toJson()}, path: ${path.contentToString()}, index: ${pathIndex + 1}")
+                    }
+                }
+                    ?: error("Field '$name' is not found because '${path[pathIndex]}' previous field is null, path: ${path.contentToString()}, index: ${pathIndex + 1}")
+
+            }
+            currentValue
+        }.onFailure {
+            throw FieldNotFoundException(
+                "Filed not found by ${path.contentToString()} path in ${toJson()} message", it
+            )
+        }.getOrThrow()
+
+        @Throws(FieldNotFoundException::class)
+        fun Message.getSimple(vararg path: String): String? = getField(*path)?.run {
+            when (kindCase) {
+                Value.KindCase.NULL_VALUE -> null
+                Value.KindCase.SIMPLE_VALUE -> simpleValue
+                else -> throw FieldNotFoundException("Value by ${path.contentToString()} path isn't simple, value: ${this.toJson()}, message: ${toJson()}")
+            }
+        }
+    }
+}
+
+class TransportMessageHolder(
+    val source: ParsedMessage,
+    val book: String,
+    val sessionGroup: String
+) : MessageHolder {
+    override val id: MessageID by lazy { source.id.toProto(book, sessionGroup) }
+    override val parentEventId: EventID? by lazy { source.eventId?.toProto() }
+    override val messageType: String
+        get() = source.type
+    override val protocol: String
+        get() = source.protocol
+    override val properties: Map<String, String>
+        get() = this.source.metadata
+    override val protoMessage: Message by lazy { source.toProto(book, sessionGroup) }
+
+    @Throws(FieldNotFoundException::class)
+    override fun getSimple(vararg path: String): String? = source.getSimple(*path)
+
+    override fun toString(): String {
+        return "TransportMessageWrapper(transport=$source, book='$book', sessionGroup='$sessionGroup')"
+    }
+
+    companion object {
+        /**
+         * Traverses the internal message and returns value by [path]
+         * @return null when the last element exist but has null value otherwise return [Any] value
+         * @throws [FieldNotFoundException] if message doesn't include full path or message structure doesn't match to path
+         */
+        @Throws(FieldNotFoundException::class)
+        fun ParsedMessage.getField(vararg path: String): Any? = runCatching {
+            require(path.isNotEmpty()) {
+                "Path to field can't be empty"
+            }
+            var currentValue: Any? = body[path.first()]
+
+            path.asSequence().drop(1).forEachIndexed { pathIndex, name ->
+                currentValue = when (currentValue) {
+                    is Map<*, *> -> (currentValue as Map<*, *>)[name]
```

Comment:
Doesn't smart cast work here? If it is because the `currentValue` is not final we can do this, I think
```kotlin
when (val current = currentValue) {
  is Map<*, *> -> current[name]
//remaining branches
}
```
Or we can do this
```kotlin
currentValue?.run {
  when (this) {
    is Map<*, *> -> get(name)
   // remaining branches
  }
}
```

Reply:
The options I showed in the comments above work fine with local variables.
```kotlin
var data: Any? = ""

data = when (val cur = data) {
    is Map<*, *> -> cur["name"]
    is List<*> -> cur.size
    else -> cur
}

data = data?.run {
    when (this) {
        is Map<*, *> -> get("name")
        is List<*> -> size
        else -> this
    }
}
```

In my opinion, it is better to have only one check and avoid explicit casting to a type to avoid mistakes in that place